### PR TITLE
Stats: build odyssey translations

### PR DESCRIPTION
**Pull request by kangzj**
#### Proposed Changes

P2: pejTkB-8l-p2

The PR adds the ability to build single language file for each language. `apps/odyssey-stats/bin/build-languages.js` is forked from `bin/build-languages.js`. The size of a single language is reduced from ~1M to ~80K max.

The overall process is:

- Build the project, where `build-tools/webpack/generate-chunks-map-plugin` generate a `chunks-map.json` for us. We use the file to merge all the strings together referenced by the built asset, for example `build.min.js`.
- Extract all strings to a pot file `odyssey-strings.pot`
- Then `apps/odyssey-stats/bin/build-languages.js` does the rest - merge all referenced strings together, pick only the necessary translations from the full translations files, and lastly write them to the target language files in `dist/languages`
- The last step is to load the built language files for Odyssey in `apps/odyssey-stats/app.jsx`

Next steps:

1. We'll automate the translation process in a follow up PR by adding the following code to `.teamcity/_self/projects/WPComPlugins.kt`:
```
	buildSteps = {
		bashNodeScript {
			name = "Translate Odyssey Stats"
			scriptContent = """
				cd apps/odyssey-stats

				# generate language files
				yarn translate
			"""
		}
	}
```

2. Another thing is, we may want to bust cache for the translations.

#### Testing Instructions

* `cd apps/odyssey-stats`
* `yarn build`
* `yarn translate`
* Ensure translations generated in `dist/languages`
* `rsync -avzP path/to/apps/odyssey-stats/dist/ sandbox:/path/to/widgets.wp.com/odyssey-stats/v1 --delete`
* Sandbox `widgets.wp.com`
* Open Stats dashboard from a Jetpack site
* Ensure Odyssey Stats loads language files properly and the interface is translated

image

#### Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: apps/odyssey-stats/.eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const { nodeConfig } = require( '@automattic/calypso-eslint-overrides' );
```
Comment (kangzj) [inline apps/odyssey-stats/.eslintrc.js:1]: Ignore files in `bin` to be linted by eslint to pass teamcity checks

**Comment by kangzj** [comment]
I noticed there're `momentjs` locale data in the language files - I'm wondering whether we could use them so that we don't have to load momentjs locale js anymore. I tried to do so with `moment.locale('fr', translation[''].momentjs_locale)`, however the dates don't look quite right to me. Anyway, we'll leave the issue to next PRs to address. Pointers would be very much appreciated, thanks.

```
...
"momentjs_locale": {
"months": [
"janvier",
"février",
"mars",
"avril",
"mai",
"juin",
"juillet",
"août",
"septembre",
"octobre",
"novembre",
"décembre"
],
"monthsShort": [
"Jan",
"Fév",
"Mar",
"Avr",
"Mai",
"Juin",
"Juil",
"Août",
"Sep",
"Oct",
"Nov",
"Déc"
],
"weekdays": [
"dimanche",
"lundi",
"mardi",
"mercredi",
"jeudi",
"vendredi",
"samedi"
],
"weekdaysShort": [
"dim",
"lun",
"mar",
"mer",
"jeu",
"ven",
"sam"
],
"weekdaysMin": [
"D",
"L",
"M",
"M",
"J",
"V",
"S"
],
"meridiem": {
"am": " ",
"pm": " ",
"AM": " ",
"PM": " "
},
"textDirection": "ltr",
"numberFormat": {
"thousands_sep": " ",
"decimal_point": ","
},
"longDateFormat": {
"LT": "HH:mm",
"L": "DD/MM/YYYY",
"LL": "D MMMM YYYY",
"LLL": "D MMMM YYYY LT",
"LLLL": "dddd, D MMMM YYYY LT"
},
"calendar": {
"sameDay": "[Today at] LT",
"nextDay": "[Tomorrow at] LT",
"nextWeek": "dddd [at] LT",
"lastDay": "[Yesterday at] LT",
"lastWeek": "[Last] dddd [at] LT",
"sameElse": "L"
},
"relativeTime": {
"future": "in %s",
"past": "%s ago",
"s": "a few seconds",
"m": "a minute",
"mm": "%d minutes",
"h": "an hour",
"hh": "%d hours",
"d": "a day",
"dd": "%d days",
"M": "a month",
"MM": "%d months",
"y": "a year",
"yy": "%d years"
},
"week": {
"dow": 0,
"doy": 4
}
}
...
```

**Review by yuliyan — APPROVED** [review]
Tested and it seems to be working as expected. LGTM!

I noticed that one part of the build script could be optimized since you're not building individual translation chunks. Consider it non-blocking for merging the PR and feel to ignore it. Here's the diff:
```diff
diff --git a/apps/odyssey-stats/bin/build-languages.js b/apps/odyssey-stats/bin/build-languages.js
index 56915e53f1..3f11c9843b 100644
--- a/apps/odyssey-stats/bin/build-languages.js
+++ b/apps/odyssey-stats/bin/build-languages.js
@@ -28,13 +28,16 @@ function createLanguagesDir() {
 	return mkdirp.sync( TEMP_PATH ) && mkdirp.sync( OUTPUT_PATH );
 }
 // Get module reference
-function getModuleReference( module ) {
-	// Rewrite module from `packages/` to match references in POT
-	if ( module.indexOf( 'packages/' ) === 0 ) {
-		return module.replace( '/dist/esm/', '/src/' ).replace( /\.\w+/, '' );
+function getModuleReference( ref ) {
+	// References need to be relative to the root of the project
+	ref = ref.replace( /^\.\.\/\.\.\//, '' );
+
+	// Rewrite ref from `packages/` to match references in POT
+	if ( ref.indexOf( 'packages/' ) === 0 ) {
+		return ref.replace( '/dist/esm/', '/src/' ).replace( /\.\w+/, '' );
 	}
 
-	return module;
+	return ref;
 }
 
 function cleanUp() {
@@ -149,39 +152,30 @@ function buildLanguages( downloadedLanguages, languageRevisions ) {
 		// CHUNKS_MAP_PATTERN is relative to the project root, while require is relative to current dir. Hence the `../`
 		const chunksMap = require( '../' + CHUNKS_MAP_PATTERN );
 
-		// Merge all source references into the main `build.min.js` chunk
-		const merged = Object.keys( chunksMap ).reduce( ( acc, key ) => {
-			acc[ 'build.min.js' ] = ( acc[ 'build.min.js' ] || [] ).concat(
-				// filepaths need to be relative to the root of the project
-				chunksMap[ key ].map( ( filepath ) => filepath.replace( /^\.\.\/\.\.\//, '' ) )
-			);
-
-			return acc;
-		}, {} );
-
-		const chunks = _.mapValues( merged, ( modules ) => {
-			const strings = new Set();
-
-			modules.forEach( ( modulePath ) => {
-				modulePath = getModuleReference( modulePath );
-				const key = /\.\w+/.test( modulePath )
-					? modulePath
-					: Object.keys( translationsByRef ).find( ( ref ) => ref.indexOf( modulePath ) === 0 );
-
-				if ( ! key ) {
-					return;
-				}
-
-				const stringsFromModule = translationsByRef[ key ] || [];
-				stringsFromModule.forEach( ( string ) => strings.add( string ) );
-			} );
-
-			return [ ...strings ];
-		} );
+		// Get only the strings that are relevant to the modules for the odyssey-stats app
+		const allModulesReferences = Object.values( chunksMap ).flat();
+		const allModulesStrings = [
+			...new Set(
+				allModulesReferences
+					.map( ( modulePath ) => {
+						modulePath = getModuleReference( modulePath );
+						const key = /\.\w+/.test( modulePath )
+							? modulePath
+							: Object.keys( translationsByRef ).find( ( ref ) => ref.indexOf( modulePath ) === 0 );
+
+						if ( ! key ) {
+							return [];
+						}
+
+						return translationsByRef[ key ] || [];
+					} )
+					.flat()
+			),
+		];
 
 		const languageRevisionsHashes = {};
 		successfullyDownloadedLanguages.forEach( ( { langSlug, languageTranslations } ) => {
-			const odysseyLanguageTranslations = _.pick( languageTranslations, chunks[ 'build.min.js' ] );
+			const odysseyLanguageTranslations = _.pick( languageTranslations, allModulesStrings );
 			odysseyLanguageTranslations[ DECIMAL_POINT_KEY ] =
 				languageTranslations[ DECIMAL_POINT_TRANSLATION ];
 			odysseyLanguageTranslations[ THOUSANDS_SEPARATOR_KEY ] =

```

**Review by jsnmoon — APPROVED** [review]
These changes test well with widgets.wp.com sandboxed on my local WordPress installation. It might be nice to add some of the context written in the body of this PR into the build script.

**Comment by kangzj** [comment]
> I noticed that one part of the build script could be optimized since you're not building individual translation chunks. @yuliyan 

Thanks for the patch. It makes sense. Applied in ed9c57e and it works well for me. 👍 

>  It might be nice to add some of the context written in the body of this PR into the build script. @jsnmoon 

Addressed in 5693500. Also thanks for the dedup.

**Review by jsnmoon — APPROVED** [review]
image

Let's ship this. I'm putting my complete trust in Yuliyan's review 🙂

**Pull request merged**